src/input/process.rs

```diff
@@ -84,6 +87,37 @@ pub fn read_processes(
     )
 }
 
+fn validate_commodities(
+    commodities: &HashMap<Rc<str>, Rc<Commodity>>,
+    flows: &HashMap<Rc<str>, Vec<ProcessFlow>>,
+) -> Result<()> {
+    for (commodity_id, commodity) in commodities {
+        if commodity.kind == CommodityType::SupplyEqualsDemand {
+            let mut has_producer = false;
+            let mut has_consumer = false;
+
+            for process_flows in flows.values() {
+                for flow in process_flows {
+                    if Rc::ptr_eq(&flow.commodity, commodity) {
+                        if flow.flow > 0.0 {
+                            has_producer = true;
+                        } else if flow.flow < 0.0 {
+                            has_consumer = true;
+                        }
+                    }
+                }
+            }
```

Comment:
You can merge the inner and outer loops here:

```suggestion
            for flow in flows.values().flatten() {
                if Rc::ptr_eq(&flow.commodity, commodity) {
                    if flow.flow > 0.0 {
                        has_producer = true;
                    } else if flow.flow < 0.0 {
                        has_consumer = true;
                    }
                }
            }
```